axelor-human-resource/src/main/java/com/axelor/apps/hr/service/AllocationLineServiceImpl.java

```diff
@@ -0,0 +1,131 @@
+public class AllocationLineServiceImpl implements AllocationLineService {
+
+  protected AllocationLineRepository allocationLineRepository;
+  protected ProjectRepository projectRepository;
+  protected EmployeeRepository employeeRepository;
+  protected PeriodRepository periodRepository;
+
+  @Inject
+  public AllocationLineServiceImpl(
+      AllocationLineRepository allocationLineRepository,
+      ProjectRepository projectRepository,
+      EmployeeRepository employeeRepository,
+      PeriodRepository periodRepository) {
+    this.allocationLineRepository = allocationLineRepository;
+    this.projectRepository = projectRepository;
+    this.employeeRepository = employeeRepository;
+    this.periodRepository = periodRepository;
+  }
+
+  @Override
+  @Transactional(rollbackOn = Exception.class)
+  public void createAllocationLine(
+      Project project, Employee employee, Period period, BigDecimal allocated) {
+    AllocationLine allocationLine = new AllocationLine();
+    allocationLine.setEmployee(employee);
+    allocationLine.setPeriod(period);
+    allocationLine.setProject(project);
+    allocationLine.setAllocated(allocated);
+    allocationLineRepository.save(allocationLine);
+  }
+
+  @Override
+  @Transactional(rollbackOn = Exception.class)
+  public void addAllocationLine(AllocationLine allocationLine) {
+    allocationLineRepository.save(allocationLineRepository.find(allocationLine.getId()));
+  }
+
+  @Override
+  @SuppressWarnings("unchecked")
+  public void addAllocationLines(Context context) {
+    if (context.get("project") == null) {
+      return;
+    }
+
+    LinkedHashMap<String, Object> projectMap =
+        (LinkedHashMap<String, Object>) context.get("project");
+    Project project = projectRepository.find(Long.parseLong(projectMap.get("id").toString()));
+
+    List<Employee> employeeList = new ArrayList<>();
+    if (context.get("employeeSet") == null) {
+      employeeList =
+          employeeRepository
+              .all()
+              .filter("self.id IN (:employeeIds)")
+              .bind("employeeIds", (List<Integer>) context.get("_employeeIds"))
+              .fetch();
+    } else {
+      for (Map<String, Object> employeeMap :
+          (Collection<Map<String, Object>>) context.get("employeeSet")) {
+        employeeList.add(employeeRepository.find(Long.parseLong(employeeMap.get("id").toString())));
+      }
+    }
+
+    List<Period> periodList = new ArrayList<>();
+    for (Map<String, Object> periodMap :
+        (Collection<Map<String, Object>>) context.get("periodSet")) {
+      periodList.add(periodRepository.find(Long.parseLong(periodMap.get("id").toString())));
+    }
+
+    BigDecimal allocated = BigDecimal.ZERO;
+    if (context.get("allocated") != null) {
+      allocated = new BigDecimal(context.get("allocated").toString());
+    }
+
+    for (Employee employee : employeeList) {
+      for (Period period : periodList) {
+        createAllocationLine(project, employee, period, allocated);
```

Comment:
Here check if an allocation line already exists for this project, employee, period and if yes, add the allocated to the existing one, else create it